<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform gene-level differential ASE analysis in the population across genes — differential_ASE_detection • ASEP</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Perform gene-level differential ASE analysis in the population across genes — differential_ASE_detection" />
<meta property="og:description" content="This function is used to detect differential ASE genes between two conditions given bulk RNA-seq data that may contain multiple genes." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ASEP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/introduction.html">Tutorial</a>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Jiaxin-Fan/ASEP">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform gene-level differential ASE analysis in the population across genes</h1>
    
    <div class="hidden name"><code>differential_ASE_detection.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function is used to detect differential ASE genes between two conditions given bulk RNA-seq data that may contain multiple genes.</p>
    </div>

    <pre class="usage"><span class='fu'>differential_ASE_detection</span>(
  <span class='no'>dat_all</span>,
  <span class='kw'>phased</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>adaptive</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>n_resample</span> <span class='kw'>=</span> <span class='fl'>10</span>^<span class='fl'>6</span>,
  <span class='kw'>parallel</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>save_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>name_out</span> <span class='kw'>=</span> <span class='kw'>NULL</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dat_all:</th>
      <td><p>bulk RNA-seq data. Must contain variables: <br /></p><ul>
<li><p>`gene`: character, gene name;</p></li>
<li><p>`id`: character, individual identifier;</p></li>
<li><p>`snp`: character, the name/chromosome location of the heterzygous genetic variants;</p></li>
<li><p>`ref`: numeric, the snp-level read counts for the reference allele if the haplotype phase of the data is unknown, and the snp-level read counts for allele aligned on the same paternal/maternal haplotype for both conditions if haplotype phase is known;</p></li>
<li><p>`total`: numeric, snp-level total read counts for both alleles;</p></li>
<li><p>`group`: character, the condition each RNA-seq sample is obtained from (i.e., pre- vs post-treatment);</p></li>
<li><p>`ref_condition`: character, the condition used as the reference for pseudo haplotype phasing;</p></li>
</ul></td>
    </tr>
    <tr>
      <th>phased:</th>
      <td><p>a logical value indicates whether the haplotype phase of the data is known or not. Default is FALSE</p></td>
    </tr>
    <tr>
      <th>n_resample:</th>
      <td><p>a numeric value indicates the maximum number of resamplings performed to obtain estimated p-value. Default is 10^6</p></td>
    </tr>
    <tr>
      <th>adaptive:</th>
      <td><p>a logical value indicates whether the resampling is done through an adaptive procedure or not. Default is TRUE <br />
By adaptive, it means first do 1000 resamplings, if the estimated p-value &lt; 0.1, increase the number of resampling, by a factor of 10, to 10^4.
if then the estimated p-value &lt; 0.01, increase the number of resampling again, by a factor of 10, to 10^5.
The procedure continuous until reaches the maximum number of resampling.</p></td>
    </tr>
    <tr>
      <th>parallel:</th>
      <td><p>a logical value indicates whether do parallel computing for the resampling precedure or not. Default is FALSE</p></td>
    </tr>
    <tr>
      <th>save_out:</th>
      <td><p>a logical value indicates whether to write out the result for each gene stepwisely to a txt file. Default is FALSE</p></td>
    </tr>
    <tr>
      <th>name_out:</th>
      <td><p>a character string indicates the output file name when save_out is set to TRUE, with the format of "XXX.txt"</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A matrix with three columns:</p><ul>
<li><p>`gene`: character, gene name;</p></li>
<li><p>`p-value`: the estimated p-value of the likelihood ratio test (LRT) statistic;</p></li>
</ul>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jiaxin Fan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


